File: 0000-0999/0227. Basic Calculator II/Solution.ts
/**
 * Author: Fbin
 * Link:https://leetcode.com/problems/basic-calculator-ii/
 * Created at: 17/03/2025
 */

function calculate(s: string): number {
  const stack: number[] = [];
  let num = 0;
  let sign = "+";
  s += "+";

  for (let c of s) {
    if (isDigit(c)) {
      num = num * 10 + Number(c);
    }

    if (!isDigit(c) && c !== " ") {
      if (sign === "+") stack.push(num);
      if (sign === "-") stack.push(-num);
      if (sign === "*") stack.push((stack.pop() || 0) * num);
      if (sign === "/") stack.push(Math.trunc(stack.pop() || 0) / num);

      num = 0;
      sign = c;
    }
  }

  return stack.reduce((acc, val) => acc + val, 0);
}

function isDigit(ch: string): boolean {
  return !isNaN(Number(ch)) && ch.trim() != "";
}

console.log(calculate(" 2+5 / 2 "));

File: 0000-0999/0227. Basic Calculator II/Solution.java
import java.util.Stack;

/**
 * Author: Fbin
 * Link:https://leetcode.com/problems/basic-calculator-ii/
 * Created at: 16/03/2025
 */

class Solution {
	public int calculate(String s) {
		Stack<Integer> st = new Stack<>();
		int num = 0;
		char sign = '+';
		s += '+';
		for (int i = 0; i < s.length(); i++) {
			char ch = s.charAt(i);
			if (Character.isDigit(ch)) {
				num = num * 10 + ch - '0';
			}

			if (!Character.isDigit(ch) && ch != ' ') {
				System.out.println("Num: " + num + '\n');
				if (sign == '+') {
					st.push(num);
				}
				if (sign == '-') {
					st.push(-num);
				}
				if (sign == '*') {
					st.push(st.pop() * num);
				}
				if (sign == '/') {
					st.push(st.pop() / num);
				}
				num = 0;
				sign = ch;
			}
		}

		int res = 0;
		while (!st.empty()) {
			res += st.pop();
		}

		return res;
	}
}
